src/morphodict/lexicon/models.py

```diff
@@ -0,0 +1,261 @@
+from __future__ import annotations
+
+import logging
+from pathlib import Path
+from typing import Dict, Literal, Union, Any
+
+from django.core.serializers.json import DjangoJSONEncoder
+from django.db import models
+from django.db.models import Q
+from django.urls import reverse
+from django.utils.functional import cached_property
+
+from CreeDictionary.API.schema import SerializedDefinition
+from CreeDictionary.utils import (
+    shared_res_dir,
+)
+
+from morphodict.analysis import RichAnalysis
+
+# How long a wordform or dictionary head can be. Not actually enforced in SQLite.
+MAX_WORDFORM_LENGTH = 60
+
+logger = logging.getLogger(__name__)
+
+
+class WordformLemmaManager(models.Manager):
+    """We are essentially always going to want the lemma
+
+    So make preselecting it the default.
+    """
+
+    def get_queryset(self):
+        return super().get_queryset().select_related("lemma")
+
+
+# This type is the int pk for a saved wordform, or (text, analysis) for an unsaved one.
+WordformKey = Union[int, tuple[str, str]]
+
+
+class DiacriticPreservingJsonEncoder(DjangoJSONEncoder):
+    """Stores Unicode strings, e.g., "pê", in the database
+
+    Instead of ASCII-fied "p\u00ea".
+    """
+
+    def __init__(self, *args, **kwargs):
+        kwargs = {**kwargs, "ensure_ascii": False}
+        super().__init__(*args, **kwargs)
+
+
+class Wordform(models.Model):
+    # Queries always do .select_related("lemma"):
+    objects = WordformLemmaManager()
+
+    text = models.CharField(max_length=MAX_WORDFORM_LENGTH)
+
+    raw_analysis = models.JSONField(null=True, encoder=DiacriticPreservingJsonEncoder)
+
+    paradigm = models.CharField(
+        max_length=MAX_WORDFORM_LENGTH,
+        null=True,
+        blank=False,
+        default=None,
+        help_text="If provided, this is the name of a static paradigm that this "
+        "wordform belongs to. This name should match the filename in "
+        "res/layouts/static/ WITHOUT the file extension.",
+    )
+
+    is_lemma = models.BooleanField(
+        default=False,
+        help_text="The wordform is chosen as lemma. This field defaults to true if according to fst the wordform is not"
+        " analyzable or it's ambiguous",
+    )
```

Comment:
I'm extremely confused. Based on _usage_, what does `is_lemma_ actually mean? Is this help text up to date?

Reply:
Oops, thank you for catching this; I had meant to update these but then forgot. You can tell something is amiss from the “blah blah blah” comment nearby.